### PR TITLE
JDK-8296995: ostream should handle snprintf(3) errors in release builds

### PR DESCRIPTION
**Pull request by tstuefe**
Small fix for a very unlikely problem. 

All streams in ostream.hpp end up using `os::snprintf()`, which uses `::vsnprintf()`. `vsnprintf(3)`can fail and return -1. 

The chance for this to happen is small. snprintf errors are usually encoding errors though not always (see third example at https://stackoverflow.com/questions/65334245/what-is-an-encoding-error-for-sprintf-that-should-return-1). I found "%ls" in one place in windows coding, so I am not sure we can always exclude the possibility of wide strings being used in our code base, or that of printing with outside-provided format strings.

In case of an error, we assert in debug builds but don't handle it in release. There, this situation gets misdiagnosed later as a buffer overflow because we cast the signedness of the result away (see `outputStream::do_vsnprintf()`). 

---

The patch is trivial. The most exciting thing is the gtest, I guess.

In release builds, we now treat this condition as an empty string write. I considered printing a clear marker into the stream instead, e.g. "ENCODING ERROR", but ultimately did not do it.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296995](https://bugs.openjdk.org/browse/JDK-8296995): ostream should handle snprintf(3) errors in release builds

### Reviewers
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**) ⚠️ Review applies to [0e1236e1](https://git.openjdk.org/jdk/pull/11160/files/0e1236e1b2c9d9358aa4480fd0750c2ebf48b362)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [086a6ec3](https://git.openjdk.org/jdk/pull/11160/files/086a6ec3ab5d22e9ae4052693aab69b34fba5bf5)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11160/head:pull/11160` \
`$ git checkout pull/11160`

Update a local copy of the PR: \
`$ git checkout pull/11160` \
`$ git pull https://git.openjdk.org/jdk pull/11160/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11160`

View PR using the GUI difftool: \
`$ git pr show -t 11160`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11160.diff">https://git.openjdk.org/jdk/pull/11160.diff</a>

</details>

### CODE REVIEW

**Review by MBaesken — APPROVED** [review]
Looks good to me.
I thought about adding some (potentially optional) log or trace in case required_len < 0 but probably it is good as it is.
Please adjust  copyright year in [test_ostream.cpp](https://github.com/openjdk/jdk/pull/11160/files#diff-c0d2076833ffa15ad831383a8ddc594763aef71e53276afeddc8bdaf1b4e887e) .

File: test/hotspot/gtest/utilities/test_ostream.cpp [hunk 1]
```diff
@@ -121,6 +122,34 @@ TEST_VM(ostream, bufferedStream_dynamic_small) {
   }
 }
 
+static void provoke_snprintf_error() {
+  const wchar_t w[] = { (wchar_t)-1, 0 };
+  char buf[20];
+  ::memset(buf, 'X', sizeof(buf));
+  int n = ::snprintf(buf + 1, sizeof(buf) - 2, "HALLO %ls", w);
+  if (n == -1) { // yes, we get an error. Retry using stringStream.
```
Comment (TheRealMDoerr) [inline test/hotspot/gtest/utilities/test_ostream.cpp:130]: Can it happen that we don't get an error, here? If so, the ASSERT below still expects one, so, the test will fail. Correct?

Comment (tstuefe) [reply test/hotspot/gtest/utilities/test_ostream.cpp:130]: Good catch. I added the death test afterwards. I'll fix it.

**Comment by tstuefe** [comment]
@TheRealMDoerr : fixed, tested on debug

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for the update! LGTM. Please update the test's Copyright header as requested by Matthias.

**Comment by tstuefe** [comment]
> Thanks for the update! LGTM. Please update the test's Copyright header as requested by Matthias.

Thanks! I missed Matthias note, thanks for reminding me.

**Comment by MBaesken** [comment]
Can't this be integrated ? What's missing?

**Pull request closed without merging**